ChiModules/LinearBoltzmannSolver/lbs_04c_io_flux_moments.cc

```diff
@@ -0,0 +1,326 @@
+    return;
+  }
+
+  //============================================= Write header
+  std::string header_info =
+    "Chi-Tech LinearBoltzmann: Flux moments file\n"
+    "Header size: 500 bytes\n"
+    "Structure(type-info):\n"
+    "size_t num_local_nodes\n"
+    "size_t num_moments\n"
+    "size_t num_groups\n"
+    "size_t num_records\n"
+    "size_t num_cells\n"
+    "Each cell:\n"
+    "  uint64_t cell_global_id\n"
+    "  size_t   num_nodes\n"
+    "  Each node:\n"
+    "    double   x_position\n"
+    "    double   y_position\n"
+    "    double   z_position\n"
+    "Each record:\n"
+    "  uint64_t     cell_global_id\n"
+    "  unsigned int node_number\n"
+    "  unsigned int moment_num\n"
+    "  unsigned int group_num\n"
+    "  double       flux_moment_value\n";
+
+  int header_size = (int)header_info.length();
+
+  char header_bytes[500];
+  memset(header_bytes, '-', 500);
+  strncpy(header_bytes, header_info.c_str(),std::min(header_size,499));
+  header_bytes[499]='\0';
+
+  file << header_bytes;
+
+  //============================================= Get relevant items
+  auto NODES_ONLY = ChiMath::UNITARY_UNKNOWN_MANAGER;
+  auto& sdm = discretization;
+  size_t num_local_nodes = discretization->GetNumLocalDOFs(NODES_ONLY);
+  size_t num_moments_t   = static_cast<size_t>(num_moments);
+  size_t num_groups      = groups.size();
+  size_t num_local_dofs  = discretization->GetNumLocalDOFs(flux_moments_uk_man);
+  size_t num_local_cells = grid->local_cells.size();
+
+  //============================================= Write num_ quantities
+  file.write((char*)&num_local_nodes,sizeof(size_t));
+  file.write((char*)&num_moments_t  ,sizeof(size_t));
+  file.write((char*)&num_groups     ,sizeof(size_t));
+  file.write((char*)&num_local_dofs ,sizeof(size_t));
+  file.write((char*)&num_local_cells,sizeof(size_t));
+
+  //============================================= Write nodal positions for
+  //                                              each cell
+  for (const auto& cell : grid->local_cells)
+  {
+    file.write((char *) &cell.global_id, sizeof(uint64_t));
+
+    size_t num_nodes      = discretization->GetCellNumNodes(cell);
+    file.write((char *) &num_nodes, sizeof(size_t));
+
+    auto   node_locations = discretization->GetCellNodeLocations(cell);
+    for (const auto& node : node_locations)
+    {
+      file.write((char *) &node.x, sizeof(double));
+      file.write((char *) &node.y, sizeof(double));
+      file.write((char *) &node.z, sizeof(double));
+    }//for node
+  }//for cell
+
+  //============================================= Write per dof data
+  try {
+    for (const auto& cell : grid->local_cells)
+      for (unsigned int i=0; i<sdm->GetCellNumNodes(cell); ++i)
+        for (unsigned int m=0; m<num_moments_t; ++m)
+          for (unsigned int g=0; g<num_groups; ++g)
+          {
+            uint64_t dof_map = sdm->MapDOFLocal(cell,i,flux_moments_uk_man,m,g);
+            double value = flux_moments.at(dof_map);
```

Comment:
at() is going to be slow, and I will argue should be used sparingly.  Instead, we should refactor this code to use [] with an assertion for bounds checking.

Reply:
Could you add more detail on exactly what you are suggesting/requesting?

Reply:
at() always performs bounds checking... something not usually done in production code.  You rarely see at() used specifically for this reason.  Instead, we should do something along the lines of:
```suggestion
            assert(dof_map < flux_moments.size());
            double value = flux_moments[dof_map];
```
Or, just leave out the assert entirely if we trust the value coming from MapDOFLocal().

Reply:
Done.